torchtitan/parallelisms/utils.py

```diff
@@ -26,3 +27,47 @@ def check_if_feature_in_pytorch(
             f"{min_nightly_version}. Please upgrade a newer version to include the "
             f"change in ({pull_request_link}) for correct {feature_name}."
         )
+
+
+def get_fully_shard_mesh_dim_names(parallel_dims: ParallelDims) -> Tuple[str, ...]:
+    """
+    Returns the names of which mesh dims that should be passed to the fully_shard()
+    function according to the `parallel_dims`. We assume that (dp_shard_enabled == True)
+    or (cp_enabled == True) is True when users call this function.
+    """
+    # Table for composable parallelisms that fully_shard() should be applied to:
+    #   parallelisms |  dp_shard_enabled|  dp_replicate_enabled|  cp_enabled|  mesh_dim_names
+    #   ------------ |  --------------- |  ------------------- |  --------- |  ---------------
+    #   no-parallel  |  False           |  False               |  False     |  error
+    #   ddp          |  False           |  True                |  False     |  error
+    #   ddp + cp     |  False           |  True                |  True      |  error
+    #   cp           |  False           |  False               |  True      |  ("cp")
+    #   hsdp         |  True            |  True                |  False     |  ("dp_replicate,  "dp_shard")
+    #   hsdp + cp    |  True            |  True                |  True      |  ("dp_replicate", "dp_cp")
+    #   fsdp         |  True            |  False               |  False     |  ("dp")
+    #   fsdp + cp    |  True            |  False               |  True      |  ("dp_cp")
+
+    if not parallel_dims.dp_shard_enabled:
+        assert parallel_dims.cp_enabled, (
+            "get_fully_shard_mesh_dim_names() should not be used when "
+            f"{'DDP' if parallel_dims.dp_replicate_enabled else 'no parallelism'} "
+            "is enabled."
+        )
+
+        if parallel_dims.dp_replicate_enabled:
+            # Composability of DDP + CP is not supported.
+            raise RuntimeError("Composability of DDP + CP is not supported.")
+
+        return ("cp",)
+    else:
+        # FSDP or HSDP, potentially with Context Parallel
+        if parallel_dims.cp_enabled:
+            if parallel_dims.dp_replicate_enabled:
+                return ("dp_replicate", "dp_cp")
+            else:
+                return ("dp_cp",)
+        else:
+            if parallel_dims.dp_replicate_enabled:
+                return ("dp_replicate", "dp_shard")
+            else:
+                return ("dp",)
```

Comment:
I somehow think these two cases can be combined as well. Please see overall comment.